java/src/main/native/src/TableJni.cpp

```diff
@@ -245,6 +248,49 @@ class native_arrow_ipc_writer_handle final {
     }
     initialized = false;
   }
+
+  std::vector<cudf::column_metadata> get_column_metadata(const table_view& tview) {
+    if (!column_names.empty() && columns_meta.empty()) {
+      // Rebuild the structure of column meta according to table schema.
+      // All the tables written by this writer should share the same schema,
+      // so build column metadata only once.
+      columns_meta.reserve(tview.num_columns());
+      size_t idx = 0;
+      for (auto itr = tview.begin(); itr < tview.end(); ++itr) {
+        columns_meta.push_back(std::move(build_one_column_meta(*itr, idx)));
+        idx ++;
+      }
+    }
+    return columns_meta;
+  }
+
+private:
+  // Still return an oject instead of being passed as an out argument, even
+  // `column_metadata` has no move constructor and would be copied.
+  cudf::column_metadata build_one_column_meta(const column_view& cview, size_t& idx) {
+    auto col_meta = cudf::column_metadata{get_column_name(idx)};
+    if (cview.type().id() == cudf::type_id::LIST) {
+      // list type requires a stub metadata for offset column, index is 0.
+      col_meta.children_meta = {{}, build_one_column_meta(cview.child(1), ++idx)};
+    } else if (cview.type().id() == cudf::type_id::STRUCT) {
+      // struct type
+      col_meta.children_meta.reserve(cview.num_children());
+      for (auto itr = cview.child_begin(); itr < cview.child_end(); ++itr) {
+        col_meta.children_meta.push_back(std::move(build_one_column_meta(*itr, ++idx)));
+      }
+    } else if (cview.type().id() == cudf::type_id::DICTIONARY32) {
+      // not supported yet in JNI, nested type?
```

Comment:
This should throw until it is supported rather than silently ignore the problem.

Reply:
good suggestion, updated